File: sdks/python/src/atap/http_client.py
"""HTTP client for the ATAP SDK with DPoP authentication."""

from __future__ import annotations

import json
from typing import Any, Dict, Optional
from urllib.parse import urlencode

import httpx

from atap.crypto import make_dpop_proof
from atap.errors import (
    ATAPAuthError,
    ATAPConflictError,
    ATAPError,
    ATAPNotFoundError,
    ATAPProblemError,
    ATAPRateLimitError,
)
from atap.models import ProblemDetail


class HTTPClient:
    """Low-level HTTP client with DPoP proof injection and error handling."""

    def __init__(
        self,
        base_url: str = "http://localhost:8080",
        timeout: float = 30.0,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._client = httpx.Client(base_url=self._base_url, timeout=timeout)

    def close(self) -> None:
        """Close the underlying HTTP client."""
        self._client.close()

    def request(
        self,
        method: str,
        path: str,
        *,
        json_body: Optional[Dict[str, Any]] = None,
        headers: Optional[Dict[str, str]] = None,
        params: Optional[Dict[str, str]] = None,
    ) -> Dict[str, Any]:
        """Make an HTTP request and return parsed JSON response."""
        h = dict(headers or {})
        if json_body is not None:
            h.setdefault("Content-Type", "application/json")

        response = self._client.request(
            method,
            path,
            json=json_body,
            headers=h,
            params=params,
        )
        return self._handle_response(response)

    def authenticated_request(
        self,
        method: str,
        path: str,
        *,
        signing_key: Any,
        access_token: str,
        platform_domain: str,
        json_body: Optional[Dict[str, Any]] = None,
        raw_body: Optional[bytes] = None,
        content_type: Optional[str] = None,
        params: Optional[Dict[str, str]] = None,
    ) -> Dict[str, Any]:
        """Make a DPoP-authenticated HTTP request."""
        url = f"https://{platform_domain}{path}"
        dpop_proof = make_dpop_proof(signing_key, method, url, access_token=access_token)

        headers: Dict[str, str] = {
            "Authorization": f"DPoP {access_token}",
            "DPoP": dpop_proof,
        }

        if content_type:
            headers["Content-Type"] = content_type

        if raw_body is not None:
            response = self._client.request(
                method,
                path,
                content=raw_body,
                headers=headers,
                params=params,
            )
        else:
            response = self._client.request(
                method,
                path,
                json=json_body,
                headers=headers,
                params=params,
            )

        return self._handle_response(response)

    def post_form(
        self,
        path: str,
        form_data: Dict[str, str],
        *,
        dpop_proof: Optional[str] = None,
    ) -> Dict[str, Any]:
        """POST form-encoded data (for OAuth token endpoint)."""
        headers: Dict[str, str] = {
            "Content-Type": "application/x-www-form-urlencoded",
        }
        if dpop_proof:
            headers["DPoP"] = dpop_proof

        response = self._client.post(
            path,
            content=urlencode(form_data).encode(),
            headers=headers,
        )
        return self._handle_response(response)

    def get_redirect(
        self,
        path: str,
        *,
        params: Optional[Dict[str, str]] = None,
        dpop_proof: Optional[str] = None,
    ) -> str:
        """GET request expecting a 302 redirect, returns the Location URL."""
        headers: Dict[str, str] = {}
        if dpop_proof:
            headers["DPoP"] = dpop_proof

        response = self._client.get(
            path,
            params=params,
            headers=headers,
            follow_redirects=False,
        )

        if response.status_code != 302:
            self._handle_response(response)
            raise ATAPError(
                f"Expected 302 redirect, got {response.status_code}",
                response.status_code,
            )

        location = response.headers.get("location", "")
        if not location:
            raise ATAPError("302 redirect with no Location header")
        return location

    def _handle_response(self, response: httpx.Response) -> Dict[str, Any]:
        """Parse response, raising typed errors for non-2xx status codes."""
        if response.status_code == 204:
            return {}

        try:
            data = response.json()
        except (json.JSONDecodeError, ValueError):
            if 200 <= response.status_code < 300:
                return {}
            raise ATAPError(
                f"HTTP {response.status_code}: {response.text}",
                response.status_code,
            )

        if 200 <= response.status_code < 300:
            return data

        # Parse RFC 7807 Problem Detail
        problem = None
        if "type" in data and "status" in data:
            problem = ProblemDetail(
                type=data.get("type", ""),
                title=data.get("title", ""),
                status=data.get("status", response.status_code),
                detail=data.get("detail"),
                instance=data.get("instance"),
            )

        status = response.status_code

        if status == 401 or status == 403:
            msg = problem.detail if problem else data.get("detail", "Authentication failed")
            raise ATAPAuthError(str(msg), status, problem)
        elif status == 404:
            msg = problem.detail if problem else "Not found"
            raise ATAPNotFoundError(str(msg), problem)
        elif status == 409:
            msg = problem.detail if problem else "Conflict"
            raise ATAPConflictError(str(msg), problem)
        elif status == 429:
            msg = problem.detail if problem else "Rate limit exceeded"
            raise ATAPRateLimitError(str(msg), problem)
        elif problem:
            raise ATAPProblemError(problem)
        else:
            raise ATAPError(
                f"HTTP {status}: {data.get('detail', data.get('message', str(data)))}",
                status,
            )

File: sdks/php/src/Http/HttpClient.php
<?php

declare(strict_types=1);

namespace Atap\Sdk\Http;

use Atap\Sdk\Crypto\DPoPProof;
use Atap\Sdk\Exception\ATAPAuthException;
use Atap\Sdk\Exception\ATAPConflictException;
use Atap\Sdk\Exception\ATAPException;
use Atap\Sdk\Exception\ATAPNotFoundException;
use Atap\Sdk\Exception\ATAPProblemException;
use Atap\Sdk\Exception\ATAPRateLimitException;
use Atap\Sdk\Model\ProblemDetail;
use GuzzleHttp\Client;
use GuzzleHttp\Exception\GuzzleException;
use GuzzleHttp\HandlerStack;
use GuzzleHttp\Psr7\Response;
use Psr\Http\Message\ResponseInterface;

/**
 * Low-level HTTP client with DPoP proof injection and error handling.
 */
class HttpClient
{
    private Client $client;
    private string $baseUrl;

    public function __construct(
        string $baseUrl = 'http://localhost:8080',
        float $timeout = 30.0,
        ?HandlerStack $handler = null,
    ) {
        $this->baseUrl = rtrim($baseUrl, '/');

        $config = [
            'base_uri' => $this->baseUrl,
            'timeout' => $timeout,
            'http_errors' => false,
        ];

        if ($handler !== null) {
            $config['handler'] = $handler;
        }

        $this->client = new Client($config);
    }

    /**
     * Get the underlying Guzzle client (for testing).
     */
    public function getClient(): Client
    {
        return $this->client;
    }

    /**
     * Make an HTTP request and return parsed JSON response.
     *
     * @param array<string, mixed>|null $jsonBody
     * @param array<string, string>|null $headers
     * @param array<string, string>|null $params
     * @return array<string, mixed>
     */
    public function request(
        string $method,
        string $path,
        ?array $jsonBody = null,
        ?array $headers = null,
        ?array $params = null,
    ): array {
        $options = [];
        $h = $headers ?? [];

        if ($jsonBody !== null) {
            $options['json'] = $jsonBody;
            $h['Content-Type'] = $h['Content-Type'] ?? 'application/json';
        }

        if (!empty($h)) {
            $options['headers'] = $h;
        }

        if ($params !== null) {
            $options['query'] = $params;
        }

        try {
            $response = $this->client->request($method, $path, $options);
        } catch (GuzzleException $e) {
            throw new ATAPException('HTTP request failed: ' . $e->getMessage(), 0, $e);
        }

        return $this->handleResponse($response);
    }

    /**
     * Make a DPoP-authenticated HTTP request.
     *
     * @param array<string, mixed>|null $jsonBody
     * @param array<string, string>|null $params
     * @return array<string, mixed>
     */
    public function authenticatedRequest(
        string $method,
        string $path,
        string $signingKey,
        string $accessToken,
        string $platformDomain,
        ?array $jsonBody = null,
        ?string $rawBody = null,
        ?string $contentType = null,
        ?array $params = null,
    ): array {
        $url = "https://{$platformDomain}{$path}";
        $dpopProof = DPoPProof::create($signingKey, $method, $url, $accessToken);

        $headers = [
            'Authorization' => "DPoP {$accessToken}",
            'DPoP' => $dpopProof,
        ];

        if ($contentType !== null) {
            $headers['Content-Type'] = $contentType;
        }

        $options = [
            'headers' => $headers,
        ];

        if ($rawBody !== null) {
            $options['body'] = $rawBody;
        } elseif ($jsonBody !== null) {
            $options['json'] = $jsonBody;
            // Remove Content-Type if using json option (Guzzle sets it)
            if (!isset($headers['Content-Type'])) {
                unset($options['headers']['Content-Type']);
            }
        }

        if ($params !== null) {
            $options['query'] = $params;
        }

        try {
            $response = $this->client->request($method, $path, $options);
        } catch (GuzzleException $e) {
            throw new ATAPException('HTTP request failed: ' . $e->getMessage(), 0, $e);
        }

        return $this->handleResponse($response);
    }

    /**
     * POST form-encoded data (for OAuth token endpoint).
     *
     * @param array<string, string> $formData
     * @return array<string, mixed>
     */
    public function postForm(
        string $path,
        array $formData,
        ?string $dpopProof = null,
    ): array {
        $headers = [
            'Content-Type' => 'application/x-www-form-urlencoded',
        ];

        if ($dpopProof !== null) {
            $headers['DPoP'] = $dpopProof;
        }

        $options = [
            'headers' => $headers,
            'body' => http_build_query($formData),
        ];

        try {
            $response = $this->client->request('POST', $path, $options);
        } catch (GuzzleException $e) {
            throw new ATAPException('HTTP request failed: ' . $e->getMessage(), 0, $e);
        }

        return $this->handleResponse($response);
    }

    /**
     * GET request expecting a 302 redirect, returns the Location URL.
     *
     * @param array<string, string>|null $params
     */
    public function getRedirect(
        string $path,
        ?array $params = null,
        ?string $dpopProof = null,
    ): string {
        $options = [
            'allow_redirects' => false,
        ];

        $headers = [];
        if ($dpopProof !== null) {
            $headers['DPoP'] = $dpopProof;
        }
        if (!empty($headers)) {
            $options['headers'] = $headers;
        }

        if ($params !== null) {
            $options['query'] = $params;
        }

        try {
            $response = $this->client->request('GET', $path, $options);
        } catch (GuzzleException $e) {
            throw new ATAPException('HTTP request failed: ' . $e->getMessage(), 0, $e);
        }

        if ($response->getStatusCode() !== 302) {
            $this->handleResponse($response);
            throw new ATAPException(
                sprintf('Expected 302 redirect, got %d', $response->getStatusCode()),
                $response->getStatusCode(),
            );
        }

        $location = $response->getHeaderLine('Location');
        if ($location === '') {
            throw new ATAPException('302 redirect with no Location header');
        }

        return $location;
    }

    /**
     * Parse response, raising typed errors for non-2xx status codes.
     *
     * @return array<string, mixed>
     */
    public function handleResponse(ResponseInterface $response): array
    {
        $status = $response->getStatusCode();

        if ($status === 204) {
            return [];
        }

        $body = (string) $response->getBody();

        try {
            $data = json_decode($body, true, 512, JSON_THROW_ON_ERROR);
        } catch (\JsonException) {
            if ($status >= 200 && $status < 300) {
                return [];
            }
            throw new ATAPException("HTTP {$status}: {$body}", $status);
        }

        if (!is_array($data)) {
            $data = [];
        }

        if ($status >= 200 && $status < 300) {
            return $data;
        }

        // Parse RFC 7807 Problem Detail
        $problem = null;
        if (isset($data['type'], $data['status'])) {
            $problem = ProblemDetail::fromArray($data);
        }

        if ($status === 401 || $status === 403) {
            $msg = $problem?->detail ?? ($data['detail'] ?? 'Authentication failed');
            throw new ATAPAuthException((string) $msg, $status, $problem);
        }

        if ($status === 404) {
            $msg = $problem?->detail ?? 'Not found';
            throw new ATAPNotFoundException((string) $msg, $problem);
        }

        if ($status === 409) {
            $msg = $problem?->detail ?? 'Conflict';
            throw new ATAPConflictException((string) $msg, $problem);
        }

        if ($status === 429) {
            $msg = $problem?->detail ?? 'Rate limit exceeded';
            throw new ATAPRateLimitException((string) $msg, $problem);
        }

        if ($problem !== null) {
            throw new ATAPProblemException($problem);
        }

        throw new ATAPException(
            sprintf('HTTP %d: %s', $status, $data['detail'] ?? $data['message'] ?? json_encode($data)),
            $status,
        );
    }
}
